### DISCUSSION TITLE
Some help to implement a TypeScript language service using the new API

### DISCUSSION THREAD

**Discussion by elchininet** [volarjs/volar.js#151] [Q&A]
Hi there,

I opened [this issue](https://github.com/volarjs/volar.js/issues/80) some months ago. I have been trying to implement a TypeScript language service with custom type defintions following [the instructions](https://github.com/volarjs/volar.js/blob/master/packages/monaco/README.md) of the new API, but I have not succeed on that.

First of all I would like to double-check these small issues and know if the fix that I applied is correct on each case:

##### 1. `createTypeScriptService` in `servicePlugins`

<img width="545" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/23320aeb-52e0-4258-9462-db065960fa64">

<img width="889" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/d0d13acb-2352-4d4e-bb55-87f759e14991">

As the `createTypeScriptService` returns an array of `ServicePlugin` it seems that it should be spreaded.

<img width="403" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/44e116c0-3003-4686-8f1d-c6fd0aadf143">

##### 2. `registerProviders` method

<img width="694" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/00df23f3-a8b2-4757-888a-8257169ddf61">

<img width="944" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/7b19633a-a537-4b45-bbe4-cbd328ac329a">

I also added the `getSyncUris` there following the types.

<img width="400" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/820a9018-2ec2-4fc5-a3e1-421df3ce3eb9">

If these fixes are OK, I would like to have your help to detect what I am missing. I have created [a minimum reproducible repo](https://github.com/elchininet/monaco-editor-with-volar).

In the [old branch](https://github.com/elchininet/monaco-editor-with-volar/tree/old) it is implemented using `@volar/monaco@1.7.9` and the language service works but without the ability of adding my own custom type definitions.

<img width="933" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/85429a79-cf64-465a-a6cd-1dd9d88f9d23">

In [the master branch](https://github.com/elchininet/monaco-editor-with-volar/tree/master) I'm using  `@volar/monaco@2.1.2` and I cannot make the language service to work, it seems that something needs to be done, but I am not able to indentify what.

<img width="506" alt="image" src="https://github.com/volarjs/volar.js/assets/3728220/76be25fa-bce5-45a7-adb8-5ca78bedefac">

On each branch just run `yarn` and `yarn start` to open the dev server.

Could you indicate what I am missing?

Regards and thanks in advance.

**Comment by remcohaszing** [comment]
You’re almost there!

You have registered the TypeScript service plugin, which adds support for TypeScript features. However, you haven’t registered any language plugins, so Volar doesn’t understand how to handle Vue files. So in `vue.worker.ts`, you should add something along the lines of:

```ts
import { createVueLanguagePlugin } from '@vue/language-core';

// …

worker.initialize((ctx) => {
  // …

  return createTypeScriptWorkerService({
    // …
    languagePlugins: [createVueLanguagePlugin(/* … */)]
  })
})
```

I don’t know the exact options `createVueLanguagePlugin` accepts, but you can have a look at https://github.com/vuejs/language-tools/blob/v2.0.6/packages/language-server/node.ts#L45-L65.

**Reply by elchininet** [reply]
Hi @remcohaszing,
Thanks for your response, I tried that before but I didn‘t get any intellisense either but good to know that the solution is indeed in that direction, I‘ll keep trying then.
Regards!

**Reply by ejfasting** [reply]
Hello @elchininet !

Did you get your minimal repo to work and/or figured out how to make things spin? :)
I've stumbled into the same issue so i'm looking for solutions

//e

**Reply by elchininet** [reply]
No, @ejfasting,
I couldn't make it work so I kept using the previous version.
If you make it work, please, share it.
Regards

**Reply by ejfasting** [reply]
@elchininet thats to bad. Thank you for letting me know :) i will share if i get something working, but im not very hopefull atm. 

Looks like the older version is the way to go, and just create the languageservice the old way :)

**Reply by ejfasting** [reply]
So, i've looked a this some more today. I've gotten to the point where the intellisense KINDA works... 
What i've found is that the initial list of completionItems are correct:
![image](https://github.com/volarjs/volar.js/assets/60758190/7169a390-f898-4987-9521-5d3c296e1eab)

Here i've even injected a custom class at the top of the virtual document, in my languagePlugin, before it is sendt to the languageService for typescript:
![image](https://github.com/volarjs/volar.js/assets/60758190/39a72ae2-e399-4114-bbd3-3bf449a8dd35)

This class can be used (even though i dont fully agree with the result here..):
![image](https://github.com/volarjs/volar.js/assets/60758190/746a816c-ec4c-4c24-96b6-d6bac72b2132)

But if i define a const, for instance, I dont get the methods that belongs to this type of variable:
![image](https://github.com/volarjs/volar.js/assets/60758190/8e1253ac-69b9-4e14-aa6a-631e770cb2e2)

I guess its something, but this just leads me to be even more confused to why it doesnt work the same through monaco as it does through a vscode extension, if it actually hits the languageserver correctly and pass it to the correct service in the same way

**Reply by remcohaszing** [reply]
Typically you have one `VirtualCode` that contains the original file, and use `embeddedCodes` to create processed virtual codes. I’m not really sure if putting processed code in the top-level `VirtualCode` is supposed to work.

Your example has one mapping, where the length exceeds the length of both the original code and the generated code from its offset. That can’t be correct. You intended to use `text.length` where you have `newText.length`.

I strongly suggest you use the [Volar Labs extension](https://volarjs.dev/core-concepts/volar-labs/) to debug your virtual code.

P.s. If you post code snippets, please use markdown code blocks, not screenshots. Screenshots of text are bad for accessibility.

**Reply by ejfasting** [reply]
Ok, so i simplified the whole example, basically removing all my custom imports and only have a document that is for the language typescript:
https://github.com/ejfasting/language-tools/blob/3e2bb689942afdaf220d46eaea627b13f7db2a82/packages/language-service/documentScanner.ts#L17 
It's still the same result, in monaco, and still works as it should in vscode. E.g. i get the initial/root list of completionItems to show, but for a defined variable it doesnt show the methods that belongs to the type. 

Regarding volar labs I'm using that when developing for vscode, but it can't debug what happens for Monaco, can it? At least I havent found a way to hook into it and see what virtual documents is being created by the monaco editor..

**Reply by ejfasting** [reply]
Hello @elchininet  
My issue seems to be that 'activateAutomaticTypeAcquisition' was mandatory to make typescript work properly:
https://github.com/ejfasting/volarjs-starter/blob/master/packages/monaco/src/custom.worker.ts#L28

**Reply by elchininet** [reply]
Hi @ejfasting,
I forgot this discussion for a while. Were you able to make it work?
Regards

>My issue seems to be that 'activateAutomaticTypeAcquisition' was mandatory to make typescript work properly:

In the minimum repo that I posted above I am using `activateAutomaticTypeAcquisition`, but it still doesn't work.

**Reply by ejfasting** [reply]
Hello @elchininet 

I set up a [sample-project](https://github.com/ejfasting/vue), where i used vue to present the monaco editor.
You can check out the sourcecode [here](https://ejfasting.github.io/vue/MonacoEditor).

It uses this [custom.worker](https://github.com/ejfasting/language-tools/blob/4adb163a68dd92f9a29f7136c2a136b65c0ae895/packages/monaco/src/custom.worker.ts), so maybe this helps (?)

//Eivind

**Reply by elchininet** [reply]
Thanks @ejfasting,
I am following [the documentation](https://github.com/volarjs/volar.js/blob/master/packages/monaco/README.md) but there are multiple breaking changes that are not reflected there and they have been shipped as minor releases, so I'll try to check your code and try to get it working from there.
Thanks again

**Reply by ejfasting** [reply]
Hi again @elchininet,

You are correct, the documentation is outdated compared to what i figured out was correct.. 
I actually ended up asking on discord. 

Hope you can make it work! :)

/Eivind

**Reply by elchininet** [reply]
>Hello @elchininet
>
>I set up a [sample-project](https://github.com/ejfasting/vue), where i used vue to present the monaco editor.
You can check out the sourcecode [here](https://ejfasting.github.io/vue/MonacoEditor).
>
>It uses this [custom.worker](https://github.com/ejfasting/language-tools/blob/4adb163a68dd92f9a29f7136c2a136b65c0ae895/packages/monaco/src/custom.worker.ts), so maybe this helps (?)

I copied exactly what you have there. For some reason I am not able to make it work, I get the same result as before: no suggestions, no intellisense, no types. Something is missing I am not able to identify what.

**Reply by elchininet** [reply]
For future visitors searching for the same, I was able to make it work partially, but at least I know that I am on the right direction. Basically I followed @remcohaszing's recommendation, but as the API of `createVueLanguagePlugin` has drastically changed, and there is no documentation about how to build it, I just ran the project using `@volar/monaco@1.7.9` which was working and debug the code to get the generated compiler options there and I manually built a `vueCompilerOptions` object. I also needed to build my own types because the types in the documentation are wrong/outdated, they just throw `TypeScript` errors, so I built them taking the parameters of the functions.

For some reason, the intellisense for the `script`, `style` and `template` tag is not there yet, but at least the types for the `defineComponent` and other variables is working.

![image](https://github.com/user-attachments/assets/0d588d6f-cb2d-4294-82ae-696f3f5afc59)

I pushed the changes to the branch [volar_monaco_2](https://github.com/elchininet/monaco-editor-with-volar/tree/volar_monaco_2). If someone find a solution to make the intellisense work completely with a `SFC` code, just create a pull rquest to the repo.

Regards

**Reply by johnsoncodehk** [reply]
@elchininet this is due to you only adding `volar-service-typescript`. You can use `getFullLanguageServicePlugins` of `@vue/language-service`, which will add all language service plugins useful for Vue.

```ts
import { createVueLanguagePlugin, resolveVueCompilerOptions } from '@vue/language-core';
import { getFullLanguageServicePlugins } from '@vue/language-service';

self.onmessage = () => {

    worker.initialize((ctx: monaco.worker.IWorkerContext) => {

        // ...

        return createTypeScriptWorkerLanguageService({
            typescript: ts,
            compilerOptions,
            uriConverter,
            workerContext: ctx,
            env,
            languagePlugins: [
                createVueLanguagePlugin(
                    ts,
                    compilerOptions,
                    resolveVueCompilerOptions({}),
                    uriConverter.asFileName
                )
            ],
            languageServicePlugins: getFullLanguageServicePlugins(ts)
        });
    });
};
```

**Reply by elchininet** [reply]
Thanks @johnsoncodehk!

That solved the intellisense for the SFC tags 🎉

![image](https://github.com/user-attachments/assets/d295491c-5cdb-41a1-9a49-04fefb00ce18)

Can I open a pull request for the [the documentation](https://github.com/volarjs/volar.js/blob/master/packages/monaco/README.md) changing all the things that I encountered that were not working following the steps?

**Reply by johnsoncodehk** [reply]
@elchininet sure, please. 🙏 (We are currently very short of manpower)

**Reply by elchininet** [reply]
@johnsoncodehk,
Perfect, I'll do it.
By the way, everything is working but my own global types, something that was working in the old version. So my custom Vue components are recognised as `any`. Does this new version require a special method to declare global types?
Regards
